vignettes/invariants.Rmd

```diff
@@ -479,28 +474,44 @@ with_df2(df2[[2]] <- NULL)
 
 `x$name <- a` and `x$"name" <- a` are equivalent to `x[["name"]] <- a`.[^column-assign-symmetry]
 
-[^column-assign-symmetry]: `$` behaves almost completely symmetrical to `[[` when comparing subsetting and subassignment.
+[^column-assign-symmetry]: `$` behaves almost completely symmetrically to `[[` when comparing subsetting and subassignment.
 
 ```{r dollar-equivalent-to-subset-assign, dftbl = TRUE}
 with_df(df$n <- 0)
-with_df(df$"n" <- 0)
 with_df(df[["n"]] <- 0)
 ```
 
-`$` does not implement partial matching but creates a new column.
+`$<-` does not perform partial matching.
 
 ```{r dollar-equivalent-to-subset-assign-pmatch, dftbl = TRUE}
 with_df(df$l <- 0)
 with_df(df[["l"]] <- 0)
 ```
 
-## Column subassignment
+## Column subassignment: `x[j] <- a`
+
+* If `j` is missing, it's replaced with `seq_along(x)`
+* If `j` is logical, it's converted to numeric with `seq_along(x)[j]`.
+
+### `a` is a list
+
+`x[j] <- a` is equivalent to `x[[j[[1]]] <- a[[1]]`, `x[[j[[2]]]] <- a[[2]]`, ....
+`a` is recycled to the same length as `j`.
+
+```{r, dftbl = TRUE}
+with_df(df[1:2] <- list(1))
+with_df(df[1:2] <- list(1, 2))
+```
 
-### Definition of `x[j] <- list(...)`.
+Only lists of length one on the RHS are recycled.
+
+```{r bracket-assign-recycles-one, dftbl = TRUE}
+with_df(df[1:2] <- list(0, 0, 0))
+with_df(df[1:3] <- list(0, 0))
+```
 
-First, if `j` is not a character vector, it is converted to numeric indices. Then, `x[j] <- a` is equivalent to performing `x[[j_1]] <- a[[1]]`, `x[[j_2]] <- a[[2]]` etc.
 When the same column is updated/removed several times, the last update wins.
-If `a` contains `NULL` values, the corresponding columns are removed *after* updating.
+If `a` contains `NULL` values, the corresponding columns are removed *after* updating. <!-- HW: why is _after_ signficant? -->
```

Comment:
The last example in line 519 shows the importance: The column `c` (index 2) is updated; if removal happened first, the `li` column would have been updated.

Should we split the "bracket-assign-def" example?

Reply:
Oh got it.